오픈마켓 II [Step 1] - zoe, aladdin

junbangg opened the pull request:
안녕하세요 그린! **🫒[@GREENOVER](https://github.com/GREENOVER)🫒**

3조 알라딘과 조이입니다

구현해야하는 기능도 많고, 고민해볼점도 많아서 PR이 조금 늦어졌어요.

수정이 필요한 부분들 한번씩 짚어주시면 감사하겠습니다 🙂

이번 리뷰도 잘 부탁드립니다🙏

## 고민했던 점

### 상품등록 후 데이터를 reload 할때, gridViewController 문제

새로운 상품을 등록한 뒤에 `gridViewController` 를 reload하려고 하면 `viewDidLoad` 가 아직 호출이 안된 상태일때는 `collectionView` 가 `nil` 이라서 런타임 오류가 발생했습니다.

<img src="https://i.imgur.com/Nw0Oi11.png" width="50%" height="50%">

그래서 그걸 방지하기 위해서 `loadViewIfNeeded` 이라는 메서드를 사용했는데 정상적으로 작동은 하는거같은데 예쁜 보라색 줄이 자꾸 생겨서 다음과 같은 방법으로 해결했습니다.

```swift
DispatchQueue.main.async {
	self.gridViewController.updateProducts(with: products)
}
```

## 조언을 얻고싶은 점

### 여러개의 API 와 통신할때는 코드 구조를 어떻게 짜면 좋은지?

하나의 앱 내부에서 다수의 API와 통신해야되는 경우가 있을것 같은데 그런 경우를 위해서 `APIService` 클래스의 구조를 어떻게 설계해야될지에 대한 고민을 했습니다.

분명히 재사용될 수 있는 코드들이 있을것 같아서 통신을 담당하는 `APIService` 와 각각 API와 통신하기 위한 책임을 추상화시킨 `protocol` 을 이용해서 구현해보려고 했는데 너무 어려워서 포기했습니다.ㅎㅎ

예를 들면

```swift
APIService(클래스) - URLSession으로 실제 통신을 해주는 역할
MarketAPIServicable(프로토콜) - MarketAPI 와 통신하기위한 기능들의 인터페이스
WeatherAPIServicable(프로토콜) - WeatherAPI 와 통신하기위한 기능들의 인터페이스
YagomAPIServicable(포로토콜) - YagomAPI 와 통신하기위한 기능들의 인터페이스
```

혹시 이런것과 관련된 best practice가 있는지, 또는 그린이 현업에서 사용해보셨거나 추천해주실만한 방법이 있는지 궁금합니다!

### APIService 의 인스턴스를 어떻게 재사용 해야되는지에 대한 고민을 해봤습니다

하나의 `APIService` 인스턴스를 여러 `ViewController` 에서 사용해야될때는 어떤식으로 재사용/전달을 해야되는지 고민을 해봤습니다.

현재 `MarketViewController` 클래스를 `SceneDelegate` 에서 생성해서 `MarketAPIService` 의 인스턴스를 주입시키고 있는데, `ProductFormViewController` 에서 다시 사용할때 똑같은 인스턴스를 전달해야되는지에 대한 고민을 해봤습니다.

그런데 현재 `MarketAPIService` 는 클래스로 구현되어있어서 `ViewController` 마다 인스턴스를 돌려쓴다면 참조 카운팅에 의한 오버헤드가 발생할 수도 있겠다?라는 생각이 들면서 고민이 더 깊어졌습니다 ㅎㅎ

그래서 차라리 필요할때만 생성해서 사용하기위해 `struct` 로 구현을 바꾼다던지, 어차피 한번만 생성해서 전역적으로 똑같이 사용할거면 싱글톤으로 구현을 해도될것 같다는 생각도 들었습니다...

방법이 너무 많은것 같고 아무리 비교해도 뭐가 맞는지 잘 모르겠어서 그린의 의견이 너무 궁금합니다!

정리를 하자면

1. `class` 로 구현
    1. 하나의 인스턴스를 생성해서 `ViewController` 마다 전달해준다
2. `struct` 로 구현
    1. 필요할때마다 인스턴스를 잠깐 생성해서 사용한다.
3. 싱글톤으로 구현해서 전역적으로 사용한다.
4. 그린이 추천해주실만한 더 좋은 방법???도 있을까요??

### DispatchQueue.main.async를 상수로 만들어주는것이 좋을까요?

- 가끔 하나의 클래스 내에서 `DispatchQueue.main` 을 여러번 접근하는 경우가 있는데, 이럴때 상수로 선언해서 재사용하는게 좋을지에 대한 고민을 해봤습니다.
- 어차피 `DispatchQueue` 의 프로퍼티를 접근하는거니까 그냥 `DispatchQueue.main` 을 그대로 사용하는게 더 직관적일까요??

## 궁금한 점

### 1. weak self에서 self를 언래핑 해야하는지

```swift
func addButtonPressed() {
        fetchPage(pageNumber: 1, itemsPerPage: 20) { [weak self] products in
            guard let products = products,
                  let self = self else { // self를 옵셔널 바인딩 하는 중
                return
            }
            self.listViewController.updateProducts(with: products)
            self.gridViewController.updateProducts(with: products)
        }
    }
```

순환참조를 막기 위해 `self` 를 weak 하게 캡처하도록 하다보니, self가 optional 타입이라서 `?`

붙여야 합니다. 이 때 `self?.listViewController.update~` 이렇게 쓰기 보단, self 를 안전하게 언래핑 해주는것이 좋다고 생각했습니다. 

그 결과 위와 같은 코드가 나왔는데 저희가 옳게 생각한 것인지 궁금합니다!

### 2. 어디에다가 놓으면 좋을까요?

따로 파일을 분리하는게 좋을까요? 

현재는 `ProductFormViewController` 에서만 쓰이기 때문에 fileprivate extension 으로 두었습니다. 

분리해주는 방법도 생각해봤는데, 프로퍼티 하나때문에 파일을 생성하면 파일이 너무 많아지고, 파일을 관리하기 어려울 것 같아서 extension으로 관리하고 있어요. 다만 ViewController의 길이가 너무 길어지는 감이 있어서 분리를 해주는게 좋을까? 하는 생각도 들었어요😅

그린은 Extension 파일을 만드는 기준이 있으신가요?

<img src="https://i.imgur.com/jYYskJy.png" width="50%" height="50%">

### 3. NameSpace의 이름

`ProductDetailViewController` 에서 사용되는 문자열들의 NameSpace를 만들었어요. 이름은 `ProductUIString` 인데 네이밍이 적절한지 궁금합니다 🙋🏻‍♀️

### 4. AlertManager의 namespace

얼럿을 관리해주는 `AlertManager` 타입이 있는데, 해당 타입을 위한 namespace를 만들지 / 말지 고민이 되었어요

둘을 고민하다가 `TitleString` / `MessageString` 타입을 만들어서 관리해주기로 했습니다.  추후 확장되었을 때 문자열들을 한 곳에서 관리하는 것이 좋을것 같다고 생각했어요. 그린은 이것에 대해서 어떻게 생각하시는지 궁금해요! 🙋🏻‍♀️

**namespace를 만들었을때** 

장점

- 문자열들을 한 곳에서 관리할 수 있다 + 반복되는 문자열을 재사용할 수 있다

단점

- 반복되는 문자열이 아니기 때문에 만들 필요가 없다. 오히려 static 프로퍼티는 사용하지 않아도 메모리에 남아있기 때문에 메모리측면에서 불리하다

namespace를 만들지 않았을 때

장점

- static property가 없기 때문에 메모리 낭비가 없다.

단점

- 문자열들을 관리하기가 어렵다

---

### 아직 구현 못한 부분

- 키보드가 textView를 가리지 않도록 하기
    - 한줄씩 쓰면 자동으로 위로 스크롤이 되는데 3줄 이상부터 알 수 없는 렉이 발생합니다..
    - 엄청 긴 텍스트를 복붙 해버리면 그건 인식이 안됩니다
- 이미지 용량 최대 300KB로 제한 하기

na-young-kwon committed:
refactor: 이니셜라이저 수정 및 StoryboardIdentifier 추가 #5

junbangg committed:
feat: updateProduct 메서드 추가 #6

- JSONParsable 포로토콜에 encode 메서드 추가

na-young-kwon committed:
feat: POST 메서드 구현 #6

- Data의 extension 추가
- registerProduct 메서드 구현
- 이미지 정보를 담기 위해 ProductImage 타입 추가

junbangg committed:
refactor: MarketAPIService 리팩토링 #6

- HTTP 메서드 네임스페이스 추가
- makeRequest 메서드 추가
- makeBody 메서드 추가

na-young-kwon committed:
test: 상품등록, 상품수정 메서드에 대한 테스트 추가 #7

- Stub request에 register/update 케이스 추가

junbangg committed:
feat: 상품등록/수정 ViewController UI 구현 #8

junbangg committed:
feat: 상품 등록 페이지 구현 #8

- fetchPage API 변경사항 적용(가격 정보 Int->Double)
- ProductDetailsViewController 구현
- 이미지 컬렉션뷰 구현

na-young-kwon committed:
feat: ProductDetailViewController 세부사항 구현 #8

- addImage 버튼을 왼쪽에 배치하도록 수정
- 네비게이션바 추가
- 제품명, 가격, 할인가격, 재고 textFiled 연결
- segmentedControl 연결
- textView에 placeholder 추가

junbangg committed:
feat: ImagePicker 및 AlertManager 구현 #8

- keyboard가 TextView가리는 현상 해결
- 이미지 개수에 대한 Alert들을 AlertManager 타입으로 구현

na-young-kwon committed:
feat: 상품등록 기능 구현 #8

- ResponseProduct 타입 추가
- Currency 타입 추가
- PostProduct 이니셜라이저 추가

junbangg committed:
feat: ProductDetailsViewController를 scrollView로 구현 #8

- ProductDetailsViewController의 addButton delegate 추가
- ListViewController reload가 제대로 안되는 상태

na-young-kwon committed:
feat: AddButtonPressedDelegate 패턴 구현 #8

- done 버튼을 눌렀을 때 fetchPage를 하도록 구현
- UITextField에 addDoneButton 메서드 추가
- 모드에 따라 ProductDetailViewController의 타이틀 변경
- 스토리보드에서 필요없어진 ViewController 삭제

junbangg committed:
refactor: 코드 전체적인 리팩토링 #8

- Namespace 타입들 추가
- 코드 컨벤션 통일
- magic number와 string literal 제거
- 너무 큰 메서드들의 로직을 분리

na-young-kwon committed:
refactor: 자잘한 리팩토링 #8

- ProductFormViewController의 StoryboardId 업데이트
- 의미없는 개행 삭제
- GridViewController의 UI관련 로직을 main에서 하도록 수정

GREENOVER commented on OpenMarket/OpenMarket/AlertManager.swift:
```diff
@@ -0,0 +1,60 @@
+    
+    private enum MessageString {
+        static let excessImages = "이미지는 5개를 초과하면 안돼요."
+        static let noImages = "이미지를 추가해주세요"
+        static let successfulRegister = "제품등록 성공"
+        static let unsuccessfulRegister = "제품등록 실패"
```
요 3개의 body 메시지는 title 메시지와 겹쳐서 사용되는것 같아요🙋🏻
title or body 메시지만 가져가도 되고 아니면 body 메시지 답게 조금 더 상세한 설명이 들어가고
title에는 "이미지 추가 필요", "성공", "실패"와 같이 간결하게 가도 괜찮을것 같습니다🤔

junbangg replied:
휴먼 에러가 발생했었네요...ㅎ

GREENOVER commented on OpenMarket/OpenMarket/AlertManager.swift:
```diff
@@ -0,0 +1,60 @@
+        static let unsuccessfulRegister = "제품등록 실패"
+    }
+    
+    static func presentExcessImagesAlert(on viewController: ProductFormViewController) {
+        let alert = UIAlertController(title: TitleString.excessImages, message: MessageString.excessImages, preferredStyle: .alert)
+        let ok = UIAlertAction(title: TitleString.ok, style: .default, handler: nil)
+        alert.addAction(ok)
+        
+        viewController.present(alert, animated: true, completion: nil)
+    }
+    
+    static func presentNoImagesAlert(on viewController: ProductFormViewController) {
+        let alert = UIAlertController(title: TitleString.noImages, message: MessageString.noImages, preferredStyle: .alert)
+        let ok = UIAlertAction(title: TitleString.ok, style: .default, handler: nil)
+        alert.addAction(ok)
+        
+        viewController.present(alert, animated: true, completion: nil)
+    }
+    
+    static func presentSuccessfulRegisterAlert(on viewController: ProductFormViewController) {
+        let alert = UIAlertController(title: TitleString.successfulRegister, message: MessageString.successfulRegister, preferredStyle: .alert)
+        let ok = UIAlertAction(title: TitleString.ok, style: .default) { _ in
+            viewController.triggerDelegateMethod()
+            viewController.dismiss(animated: true, completion: nil)
+        }
+        alert.addAction(ok)
+        
+        viewController.present(alert, animated: true, completion: nil)
+    }
+    
+    static func presentUnsuccessfulRegisterAlert(on viewController: ProductFormViewController) {
+        let alert = UIAlertController(title: TitleString.unsuccessfulRegister, message: MessageString.unsuccessfulRegister, preferredStyle: .alert)
+        let ok = UIAlertAction(title: TitleString.ok, style: .default)
+        alert.addAction(ok)
+        
+        viewController.present(alert, animated: true, completion: nil)
+    }
```
이렇게 메서드로 노출되는 얼럿에 대한 케이스들을 다 만들수도 있지만
그렇게 되면 추후 너무 확장성이 떨어질것 같아요..ㅠ
(되게되게 많아지면 매번 추가해줘야하고 휴먼 에러를 발생시키기 너무 좋아보입니다.)
그래서 저는 주로 사용하는 방식은 `presentAlert`이라는 통합 메서드로 만들고 파라미터로 지금으로 적용시키면
VC, title, message, completion을 받아 구현해주고 호출부에서 이 파라미터들을 넘겨주는게 더 안전하다고 생각해요🙋🏻
최대한 여기서는 추상화를 시키고 호출하는쪽에서 메서드 파라미터 메시지를 통해 소통하는 방식이 더 낫지 않을까합니다!

junbangg replied:
저희도 그런 방법으로 리팩토링을 해봤는데 코드의 양도 줄어들고 좋은것 같아요!

GREENOVER commented on OpenMarket/OpenMarket/Extensions/UITextField+Extensions.swift:
```diff
@@ -0,0 +1,36 @@
+
+import UIKit
+
+extension UITextField {
```
저는 이렇게 기본 타입에 대해서도 별도 파일로 분리해주는거를 선호합니다!

na-young-kwon replied:
TextField / TextView 모두 별도의 파일로 분리해주었습니다!

GREENOVER commented on OpenMarket/OpenMarket/Extensions/UITextField+Extensions.swift:
```diff
@@ -0,0 +1,36 @@
+import UIKit
+
+extension UITextField {
+    func addDoneButton() {
```
오호! 툴바형태로 직접 텍스트필드의 완료 기능을 만들어주셨군요!
정상작동하는것 확인했습니다ㅎㅎ
그런데 하나 궁금한게 있습니다..!
보통 저런 툴바도 있으면서 동시에 아래와 같은 리턴 키로도 다음 텍스트 필드로 becomeFirstResponder되거나 아니면 현재와 같이 resignFirstResponder되거나 하는데 이 기능은 빠져 있어서 고려해보시면 좋을것 같아요ㅎㅎ
![스크린샷 2022-01-25 오전 9 45 12](https://user-images.githubusercontent.com/72292617/150889528-08b4eaf7-058e-4d03-b0bc-3ea5111aa815.png)

junbangg replied:
중간에 가격 관련된 키보드들은 Number Pad 로 되어있어서 리턴 버튼이 없더라고요ㅜㅜ
그래서 구현해놓은 완료 버튼으로 다음 responder 찾는 방법으로 바꾸고 리턴 버튼으로는 가능한것들만 구현해봤습니다!

GREENOVER replied:
그러면 혹시 넘버 패드는 만들어두신 "확인"을 눌렀을때 hide되면서 다음 키보드로 넘어가는 구현은 적용이 적절치 않을까요?
문득 생각나서 몰라서 여쭤봅니다..!🙋🏻

na-young-kwon replied:
<img src="https://i.imgur.com/bYD7MN8.gif" width="40%" height="40%">
이렇게 수정해보았는데, 이 기능을 말씀하시는게 맞으시죠?!

GREENOVER replied:
와우!! 네 맞습니다!
제가 말씀드린거와 정확히 일치하네요!
그런데 두분은 이 기능에 대해 괜찮다고 생각하시나요?

na-young-kwon replied:
네! 완전 필요한 기능인 것 같아요 
앱 사용할때는 인지하지 못했는데
이런 사소한곳에도 개발자의 수고가 녹아있는지 이번에 알게되었어요ㅋㅋㅋㅋ

GREENOVER commented on OpenMarket/OpenMarket/Extensions/UITextField+Extensions.swift:
```diff
@@ -0,0 +1,36 @@
+        inputAccessoryView = doneToolBar
+    }
+    
+    @objc
+    private func dismissKeyboard() {
+        resignFirstResponder()
+    }
```
오 resignFirstResponder로 구현하시고 아주 좋네요!
하나 더 욕심내보자면 현재 기능엔 없는데 완료나 타 텍스트필드 말고 다른 뷰 영역 (빈 영역)을 클릭하거나 했을때
키보드가 내려가는 기능이 있다면 더 사용자 경험에 좋을것 같습니다ㅎㅎ

junbangg replied:
이런 디테일함 너무 좋은것 같네요:) 👏👏  `UIViewController` extension으로 구현해봤습니다!

GREENOVER commented on OpenMarket/OpenMarket/GridViewController.swift:
```diff
@@ -24,7 +24,8 @@ final class GridViewController: UIViewController {
 
     required init?(coder: NSCoder) {
-        fatalError("class does not support nscoder")
+        assertionFailure("init(coder:) has not been implemented")
+        return nil
```
👍👍👍

GREENOVER commented on OpenMarket/OpenMarket/GridViewController.swift:
```diff
@@ -37,6 +38,17 @@ final class GridViewController: UIViewController {
+    
+    // MARK: - Internal Methods
+ 
+    func updateProducts(with products: [Product]) {
+        self.products = products
+        
+        DispatchQueue.main.async {
+            self.loadViewIfNeeded()
+            self.collectionView.reloadData()
+        }
+    }
```
업데이트는 되지만 아직도 20개의 post밖에 못가져와서 아쉽네요😭

GREENOVER commented on OpenMarket/OpenMarket/GridViewController.swift:
```diff
@@ -56,11 +68,15 @@ extension GridViewController {
 
 extension GridViewController: UICollectionViewDataSource {
-    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
+    func collectionView(_ collectionView: UICollectionView,
+                        numberOfItemsInSection section: Int
+    ) -> Int {
```
전체적인 코드 컨벤션을 따른다면 아래와 같이 변경되면 맞을것 같아요🙋🏻
```swift
func collectionView(
    _ collectionView: UICollectionView,
    numberOfItemsInSection section: Int
) -> Int {
....
}
```
아래 컨벤션도 비슷하게 적용해볼 수 있겠습니다🙋🏻

junbangg replied:
이게 아직도 적응이 잘 안돼서 매번 방법을 까먹네요..고쳤습니다!

GREENOVER commented on OpenMarket/OpenMarket/MarketViewController.swift:
```diff
@@ -74,6 +82,20 @@ extension MarketViewController {
+                ProductFormViewController(delegate: self, pageMode: .register, coder: coder)
+            }
+        ) else {
+            fatalError("실패")
```
이 부분도 fatalError대신 얼럿을 노출해주고 후속 처리를 해주면 어떨까요?🙋🏻
저는 코드에서 fatalError를 안쓰려하는 스타일을 가지고 있습니다🤔

junbangg replied:
이부분은 저희가 리팩토링하는걸 놓쳤네요..바꿨습니다! 감사합니다~

GREENOVER commented on OpenMarket/OpenMarket/MarketViewController.swift:
```diff
@@ -118,18 +140,40 @@ extension MarketViewController {
-        apiService?.fetchPage(pageNumber: pageNumber, itemsPerPage: itemsPerPage) { [weak self] result in
-            self?.stopLoadingIndicator()
+        apiService?.fetchPage(pageNumber: pageNumber,
+                              itemsPerPage: itemsPerPage) { [weak self] result in
```
요기두 파라미터 컨벤션을 다른곳과 맞춰볼 수 있을것 같습니다🙋🏻

na-young-kwon replied:
이부분 포함 코드 전체적으로 띄어쓰기 컨벤션 통일해주었습니다!

GREENOVER commented on OpenMarket/OpenMarket/MarketViewController.swift:
```diff
@@ -118,18 +140,40 @@ extension MarketViewController {
+
+extension MarketViewController: AddButtonPressedDelegate {
+    func addButtonPressed() {
+        fetchPage(pageNumber: 1, itemsPerPage: 20) { [weak self] products in
```
음.. 상품 등록을 하는 시점에서 얼럿 노출되고 ok 버튼을 누르면 취해지는 액션 호출에 따라 이 부분이 호출되는것 같은데
해당하는 메서드가 해야하는 일이 많으면서도 네이밍만으로 로직을 유추하기가 힘듭니다😭
상품페이지 fetch되는것과 list, grid에 업데이트 치는것을 별도의 메서드로 만들어 호출하는게 더 이해하기 쉽지 않을까해요🤔
정리하자면 한 메서드에서 많은 역할을 떠안고 있어보입니다.
또 네이밍만 보면 얼럿의 확인 버튼을 눌렀을때가 아닌 상품등록 버튼을 눌렀을때 이 로직들이 이뤄진다고 보이는데
사실상으론 얼럿의 ok 액션의 핸들러로 호출된다고 보여서요🤔
조금 더 괜찮은 리팩토링이 있을것 같아요🙋🏻

GREENOVER commented on OpenMarket/OpenMarket/Model/MarketAPIService.swift:
```diff
@@ -106,6 +136,42 @@ extension MarketAPIService: APIServicable {
+        
+        return body
+    }
+    
```
Request, body를 별도로 공통되게 분리한것 좋네요!👍

GREENOVER commented on OpenMarket/OpenMarket/Model/MarketAPIService.swift:
```diff
@@ -106,6 +136,42 @@ extension MarketAPIService: APIServicable {
+        body.append(convertFormField(fieldName: "params", json: jsonData, using: boundary))
+        
+        for image in images {
+            let data =  convertFileData(fieldName: "images", fileName: image.fileName, mimeType: "image/\(image.type.description)", fileData: image.data!, using: boundary)
```
여기 파라미터 구현부가 좀 긴것 같아서 가독이 쪼~금 떨어져요🥲
좀 더 가독성 있게 바꿔볼 수 있을것 같습니다🙋🏻

na-young-kwon replied:
수정해주었습니다!
추가로 강제언래핑 하는 부분을 개선해보았어요 :)

GREENOVER commented on OpenMarket/OpenMarket/Model/MarketAPIService.swift:
```diff
@@ -134,6 +200,42 @@ extension MarketAPIService {
 }
 
 
+
```
불필요한 줄바꿈이 들어간것 같습니다🙋🏻

na-young-kwon replied:
앗 이부분을 놓쳤네요! 수정완료했습니다 :)

GREENOVER commented on OpenMarket/OpenMarket/ProductFormViewController.swift:
```diff
@@ -0,0 +1,394 @@
+    }()
+    
+    private lazy var imagePicker: UIImagePickerController = {
+        let imagePicker = UIImagePickerController()
```
PHPicker라는 친구도 있는데 어떤 차이가 있을까요?
UIImagePicker로 사용하신 이유가 궁금합니다🙋🏻

junbangg replied:
저희는 그런게 있는줄을 몰랐습니다! 찾아보니까 조금 더 다양한 기능들이 제공되는것 같더라고요!
조금 더 자세히 공부를 해보겠습니다 :)

GREENOVER commented on OpenMarket/OpenMarket/ProductFormViewController.swift:
```diff
@@ -0,0 +1,394 @@
+// MARK: - AddButtonPressedDelegate Protocol
+
+protocol AddButtonPressedDelegate: AnyObject {
+    func addButtonPressed()
```
네이밍을 조금 바꿔보면 어떨까요?
addDoneButton과 해당하는 네이밍과 조금 겹쳐서 약간 이해하기 어려웠습니다ㅠ
예를들어 요건 상품등록 버튼 동작에 해당하는거니
`registerButtonTapped`같은..?

na-young-kwon replied:
오 그럴수도 있겠네요,, 수정완료하였습니다
추가로 통일성을 위해 버튼이 눌렸을 때 네이밍을 `pressed` -> `tapped` 로 변경해주었습니다!

GREENOVER commented on OpenMarket/OpenMarket/ProductFormViewController.swift:
```diff
@@ -0,0 +1,394 @@
+    override func viewWillAppear(_ animated: Bool) {
+        super.viewWillAppear(true)
+        
+        productNameTextField.becomeFirstResponder()
```
뷰 노출 후 상품명에 포커싱되도록 구현하것 너무 좋네요👍

GREENOVER commented on OpenMarket/OpenMarket/ProductFormViewController.swift:
```diff
@@ -0,0 +1,394 @@
+        }
+        registerProduct()
+        AlertManager.presentSuccessfulRegisterAlert(on: self)
+    }
```
오 구현 좋네요!
그런데 현재는 이미지 누락이 아니고 다른 필드가 누락되면 상세히 어떤 필드가 필요한지 누락되었는지에 대한 얼럿이 아닌
통합적인 제품 등록 실패 얼럿으로 되는것 같습니다🤔
문제는 없지만 상세하면 더 좋지 않을까? 라는 의견이 있습니다🙋🏻

junbangg replied:
이 부분에 대해서 고민을 하다가 각 `UITextField` 마다 입력값에 따라 border의 색을 바꿔주는 방법으로 구현을 해봤습니다!

na-young-kwon replied:
<img src="https://i.imgur.com/1xkmGAb.gif" width="40%" height="40%">

이렇게요!

GREENOVER replied:
오..! 멋있어졌네요ㅎㅎ
그런데 정말 제가 아무것도 모르는 사용자라면 저 초록 border가 어떤 역할을 해주는지 감을 못잡을 수도 있을것 같아요.
또한 해당 필드(넘버 필드들)에 숫자를 넣었다가 삭제하고 다른 필드를 클릭하면 0으로 의도치않은 값이 자동으로 들어가면서 
border가 초록으로 바뀌는데 의도된건가요..?
이 방향으로 가신다면 조금 더 보완이 되면 좋긴할것 같습니다!
만약 상품가격까지만 진행하고 Done을 누르면 나머지 필드들(입력이 안된 필드)을 Red Border를 변경하거나 하는 작업이 있다던가하면
더 좋겠어요🤔

GREENOVER commented on OpenMarket/OpenMarket/ProductFormViewController.swift:
```diff
@@ -0,0 +1,394 @@
+        guard let postProduct = makePostProduct() else {
+            AlertManager.presentUnsuccessfulRegisterAlert(on: self)
+            return
+        }
```
여기서 필드가 누락된 상품들을 순차적으로 얼럿을 보여주면 더 구체적으로 왜 제품 등록에 실패했는지 파악하기
좋은 얼럿이 되지 않을까 싶습니다..!🙋🏻

GREENOVER commented on OpenMarket/OpenMarket/ProductFormViewController.swift:
```diff
@@ -0,0 +1,394 @@
+            image: image
+        )
+        
+        images.append(image)
```
images에 삽입 전 크기에 따라 리사이즈가 필요해보이네요🤔
아마 많이 찾아보셨겠지만..!?
https://www.createwithswift.com/uiimage-resize-resizing-an-uiimage/
이런 레퍼런스도 잘 정리되어 있는것 같습니다🙋🏻

na-young-kwon replied:
이런방법으로 리사이징을 하는군요! 
extension으로 기능 추가해주었고, 이것에 대해 좀 더 공부해봐야 겠어요!

junbangg replied:
감사합니다! 이것도 반영 했습니다~~

GREENOVER reviewed (changes requested):
@junbangg @na-young-kwon 
두분 상품등록 잘 구현해주셨네요!
고생 많으셨습니다 ㅎㅎ
이전 스텝부터 차근차근 잘 해오셨고 구조도 잘 나눠주셔서 가독도 좋고 리뷰드리기 편했씁니다👍
몇가지 추가 및 수정될 사항들도 있는거 포함해서 리뷰 해보겠습니다🙌
***
### 칭찬 드리고 싶은 부분
- 전체적으로 파일 구조를 너무 잘 나눠주셔서 편했습니다 💯
- 텍스트필드에 대해 resign, become 개념을 잘 사용하셨네요 💯
- 뷰와 뷰컨을 명확히 구분해주셔서 방대한 코드양을 줄인것 같아요 💯
- 마크 주석을 잘 활용하셔서 더욱 로직 흐름 파악에 편했습니다 💯
***
### 조언이 필요한 부분에 대한 저의 의견
1. 여러개의 API 와 통신할때는 코드 구조를 어떻게 짜면 좋은지?
- 음 마지막 맥락에 말씀하신것과 같이 다양한 API에 대해 프로토콜로 작성하는걸 저는 하고 있습니다!
이를 통해 Request, Response의 각기의 타입들도 구현해서 사용한다면 더욱 공통적으로 묶을 수 있을것 같아요.
어렵긴 하겠지만 고민하셨던 결론을 저도 추천하는 바입니다🙌
2. APIService 의 인스턴스를 어떻게 재사용 해야되는지에 대한 고민을 해봤습니다
- 저도 이와 고민이 좀 많았었는데 struct로 만들어 사용합니다!
이는 고민하셨던것과 같이 클래스로하거나 싱글턴으로 간다면 통신하고 메모리 해제에 대해 신경써줘야하는 부분때문입니다.
크게 class로 했을때의 이점보다 메모리에 대해서 덜 신경쓰는게 효율적이라 생각했기에 struct로 구성해주는 편입니다🙋🏻
3. DispatchQueue.main.async를 상수로 만들어주는것이 좋을까요?
- 결론은 저는 별도 상수로 만들 필요까진 없을것 같아요.
나중에 전체적인 클래스에 대해 의존성 주입 개념을 도입한다면 상위 클래스로부터 DispatchQueue.main을 받아와 init해주어 사용된다면
이것또한 상수 처리와 비슷하게 동작한다고 보입니다.
지금 구현으로 충분하다고 생각합니다!
***
### 궁금하신 부분에 대한 저의 의견
1. weak self에서 self를 언래핑 해야하는지
- 네! 언래핑하는게 제일 안전하긴 하겠네요ㅎㅎ
지금 구현 좋습니다!
2. 어디에다가 놓으면 좋을까요?
- 저는 혹시 모를 확장성과 유지보수를 위해 기본 타입들에 대한 Extension들은 현재는 그 부분에서만 쓰이더라도
파일로 분리하고 public하게 가져가는 스타일을 추구하고 있습니다ㅎㅎ
파일로 잘 나눠주셨고 scope에 대해 고민인데 애매할때는 public으로 가는것도 방법일것 같아요!
3. NameSpace의 이름
- 사실 저도 두분의 요 부분 코드를 보고 네이밍에 대해 고민을 좀 해봤습니다.
별도 코드 리뷰 코멘트로는 남기지 않았습니다.
그 이유는 딱히 적절하지 않다고 생각이 들진 않았고 조금 더 괜찮은 네이밍이 있을것 같긴했는데 생각이 나지 않았어요🥲
애매했던건 조금 포괄적인 느낌이여서 그랬던것 같습니다.
어떤 UI의 String인지에 구체적이면 어떨까라는 의견도 있습니다🤔
예를들어 ProductTextFieldString 같은 느낌..?
4. AlertManager의 namespace
- 네임 스페이스로 만든건 스타일 차이이기도 하지만 명확해서 저는 좋았습니다..!
다만 확장성과 유지보수 측면에서 고려해본다면 나중에 해당 파일에 대한 관리가 너무 많아질것 같아요ㅠ
매번 추가하고 정의하다가 문자열 관리에 대해서는 정확하지만 다른 부분에서 휴먼 에러가 나지 않을까 생각이 들었어요..!
리뷰에도 남기긴 했지만 저는 얼럿에 대해서는 namespace 관리는 하지 않는 편입니다🤔
파라미터로 title, message, handler등을 받는 공통 alert만 정의하고 호출부에서 넘겨줘 쓰는 편입니다!
이건 스타일 차이니니 정말..ㅎㅎ 이에 대해 코멘트로 조금 자세히 기록해봤으니 봐주시면 감사하겠습니다🙇🏻
***
### 미완료된 구현에 대한 저의 의견
1. TextField & Keyboard
- 이 부분에 대해서는 확인했는데 저도 아직 파악이 잘 안되네요🥲
100프로 발생하진 않고 복붙했다 다른 UI 클릭이나 시간지나고 다시 누르면 다시 스크롤링되어 포커싱되는것도 좀 애매하네요🤔
특정조건에서 스크린의 사이즈에 대해 포커싱이 잘 이뤄지지 않는것 같긴한데..
요 부분은 디버깅을 해보며 더 고민해봐야할것 같습니다!
2. 이미지 resize
- 요 부분은 추가로 구현이 필요해보입니다..!
리뷰 코멘트로 관련 레퍼런스를 남겨보긴 했는데 해당 부분에 resize를 하는 로직만 추가하면 해결될것으로 보입니다🤔
***
### 보완되었으면 좋을 부분
- 현재 상품등록 시 상세설명에 대해서만 TextField 작성하다 다른 필드를 클릭 후 다시 재 작성을 이어가려고 클릭하면
작성되어 있는 text들이 초기화 되어 버리는 문제가 있습니다.
요 부분이 위 textField 복붙이나 포커싱 그 문제와 연관이 있는지는 모르겠지만 수정이 필요해 보였습니다🙋🏻
- 이건 마이너하긴 하지만 보통 텍스트필드에서 다음 텍스트필드가 있을경우 완료나 return 키를 누르면 다음 텍스트필드에 포커싱되는 즉
현재 필드는 resign하고 다음 필드에 대해 become되도록 로직을 설정하고 있습니다.
정말 스타일이긴 하지만.. 여유가 되신다면 이런 디테일이 포함되면 좋겠습니다😀
- 상품등록 과정에서 이미지에 대해 현재 등록만되고 등록된 이미지를 제거할 수 없는데 이부분은 추가 구현이 필요해보입니다🤔
등록 화면에서는 등록된 이미지를 클릭하거나 혹은 특정한 이미지에 버튼을 overlay하여 버튼 클릭 시 해당 이미지를 제거할 수 있는 구현이 있으면 좋을것 같아요😀
***
조금 의견이 길어졌네요🥲
조금 다듬을 부분이 있다고 판단해 Request Changes로 해놓겠습니다..!
수정 후 리뷰 재요청을 해주시면 감사하겠습니다ㅎㅎ

junbangg committed:
refactor: 등록된 이미지를 제거할 수 있는 기능 추가 #8

- protocol 메서드명 변경 (buttonPressed -> buttonTapped)
- AddImageButtonCell, ImageCell rounded corners 추가

na-young-kwon committed:
feat: 완료버튼 누르면 다음 텍스트필드가 포커싱되는 기능 구현 #8

- 텍스트필드 외의 부분을 누르면 키보드가 사라지도록 구현
- TextView Extensions 파일 추가
- 전체적인 띄어쓰기 컨벤션 수정

junbangg committed:
refactor: AlerManager 기능을 하나의 extension 메서드로 수정 #8

- UIImage의 사이즈를 조절하는 extension 메서드 추가
- 가독성이 떨어지는 메서드 파라미터에 적절한 줄바꿈 추가

na-young-kwon committed:
feat: 누락된 필드를 알려주는 기능 추가#8

- textField / textView에 입력될 때 마다 색깔로 성공/실패 알림
- addTextFieldTarget 메서드 추가
- changeTextFieldBorder 에서 border 색을 바꾸도록 구현

na-young-kwon commented:
### 추가로 궁금한 점
self의 사용

가독성을 위해 불필요한 `self` 는 쓰지 않는 것이 좋다고 생각합니다. 
꼭 필요한 상황(클로저에서 / 파라미터 명과 프로퍼티 이름이 같을 때) 에서 만 쓰는 것이 좋다고 알고 있어요.

이 때 extension에서의 self의 사용도 이에 해당하는지 궁금합니다

1번 에서는 self를 꼭 써야하지만,
2번에서는 self를 생략해주어도 컴파일 에러는 나지 않는 상황입니다.

2번 상황에서도 self를 명시적으로 적어 줄 필요가 있는지 궁금해요🙋🏻‍♀️
이런 고민을 한 이유가 `self` 를 적지 않으면 뭔가 허전한 느낌이 나기 때문이에요,, (2-1, 2-2 둘 다)

1

```swift
extension Int {
    var stringFormat: String {
        return String(self)
    }
}
```

2-1

```swift
extension Data {
    mutating func append(_ string: String) {
        guard let data = string.data(using: .utf8) else {
            return
        }
        self.append(data)
    }
}
```

2-2
self를 적지 않으면 어떤 text를 의미하는지 알 수 없을것같은 느낌이 들어요
```swift
extension UITextView {
    var isValid: Bool {
        return self.text != ProductUIString.defaultDescription
    }
}
```

그린은 `self` 키워드를 명시하는 기준이 있으신가요?

GREENOVER commented on OpenMarket/OpenMarket/Extensions/UIViewController+Extensions.swift:
```diff
@@ -16,4 +16,16 @@ extension UIViewController {
+        handler: ((UIAlertAction) -> ())?
+    ) {
+        let alert = UIAlertController(title: alertTitle, message: alertMessage, preferredStyle: .alert)
+        let ok = UIAlertAction(title: "확인", style: .default, handler: handler)
```
추후 어떤 얼럿에서는 이 ok 버튼이 "확인"이 아닌 다른 텍스트로 교체되면 하나 더 만들어줘야할것 같은데
이 title도 기본적으로 이니셜 시 받아올 수 있도록 해주고 기본값으로는 "확인"을 넣어주면 더 좋을것 같아요ㅎㅎ

GREENOVER reviewed (approved):
고생하셨습니다!ㅎㅎ
더 보완되고 요구사항도 완벽히 만족시켰네요!
재리뷰를 단 부분이 2개 정도 있습니다..!
그리고 마이너하지만 스크롤 시 위 Navigation 영역을 가려버리는 현상이 있어 제보드립니다🥲
![스크린샷 2022-01-26 오후 5 48 09](https://user-images.githubusercontent.com/72292617/151132437-bbb7dd0e-d12e-47a2-84f1-2eaca619b2a8.png)
***
### 추가 질문에 대한 저의 의견
- 우선 저는 적용되는 부분이 애매하다면 self를 왠만하면 써주긴합니다.
이것도 권장되는게 있는지 저도 잘 몰랐네요🥲
익스텐션해서 만들어줄때 text를 예를들어 같은 네이밍으로 text를 받게되는데 이럴때는 text라는 동일 네이밍이 겹쳐
어디의 text인지 애매모호한 경우라고 판단되요.
이럴때 self를 안붙여도 되지만 더 명확하기에 저는 앞서 말씀드린것처럼 왠만해선 붙이긴해요..!
***
다시 한번 수고하셨어요..!
보완되어야 할 점이 있다면 이후 스텝에서 완성시켜보면 좋을것 같아요ㅎㅎ
머지하겠습니다🙌
